code/game/machinery/computer/launchpad_control.dm

```diff
@@ -12,6 +12,107 @@
+
+	var/datum/port/input/launchpad_id
+	var/datum/port/input/x_pos
+	var/datum/port/input/y_pos
+	var/datum/port/input/send_trigger
+	var/datum/port/input/retrieve_trigger
+
+	var/datum/port/output/sent
+	var/datum/port/output/retrieved
+	var/datum/port/output/on_fail
+	var/datum/port/output/why_fail
+
+	var/obj/machinery/computer/launchpad/attached_console
+
+/obj/item/circuit_component/bluespace_launchpad/Initialize()
+	. = ..()
+	launchpad_id = add_input_port("Launchpad ID", PORT_TYPE_NUMBER, FALSE)
+	x_pos = add_input_port("X offset", PORT_TYPE_NUMBER, FALSE)
+	y_pos = add_input_port("Y offset", PORT_TYPE_NUMBER, FALSE)
+	send_trigger = add_input_port("Send", PORT_TYPE_SIGNAL)
+	retrieve_trigger = add_input_port("Retrieve", PORT_TYPE_SIGNAL)
+
+	sent = add_output_port("Sent", PORT_TYPE_SIGNAL)
+	retrieved = add_output_port("Retrieved", PORT_TYPE_SIGNAL)
+	why_fail = add_output_port("Fail reason", PORT_TYPE_STRING)
+	on_fail = add_output_port("Failed", PORT_TYPE_SIGNAL)
+
+/obj/item/circuit_component/bluespace_launchpad/Destroy()
+	launchpad_id = null
+	x_pos = null
+	y_pos = null
+	send_trigger = null
+	retrieve_trigger = null
+	sent = null
+	retrieved = null
+	on_fail = null
+	why_fail = null
+	return ..()
+
+/obj/item/circuit_component/bluespace_launchpad/register_usb_parent(atom/movable/parent)
+	. = ..()
+	if(istype(parent, /obj/machinery/computer/launchpad))
+		attached_console = parent
+
+/obj/item/circuit_component/bluespace_launchpad/unregister_usb_parent(atom/movable/parent)
+	attached_console = null
+	return ..()
+
+/obj/item/circuit_component/bluespace_launchpad/input_received(datum/port/input/port)
+	. = ..()
+
+	if(.)
+		return
+
+	if(!attached_console || length(attached_console.launchpads) == 0)
+		why_fail.set_output("No launchpads connected!")
+		on_fail.set_output(COMPONENT_SIGNAL)
+		return
+
+	var/current_launchpad = launchpad_id.input_value
+	if(isnull(current_launchpad))
+		return
+
+	var/obj/machinery/launchpad/the_pad = attached_console.launchpads[current_launchpad]
+	if(isnull(the_pad))
+		return
+
+	var/x_dest = x_pos.input_value
+	if(isnull(x_dest))
+		return
+
+	var/y_dest = y_pos.input_value
+	if(isnull(y_dest))
+		return
+
+	if(x_dest > the_pad.range || y_dest > the_pad.range)
+		why_fail.set_output("Cannot go that far! Current maximum reach: [the_pad.range]")
+		on_fail.set_output(COMPONENT_SIGNAL)
+		return
+
+	var/checks = attached_console.teleport_checks(the_pad)
+
+	if(!isnull(checks))
+		why_fail.set_output(checks)
+		on_fail.set_output(COMPONENT_SIGNAL)
+		return
+
+	the_pad.set_offset(x_dest, y_dest)
+
+	if(COMPONENT_TRIGGERED_BY(send_trigger, port))
+		the_pad.doteleport(null, TRUE, inserter = parent.get_creator())
+
+	if(COMPONENT_TRIGGERED_BY(retrieve_trigger, port))
+		the_pad.doteleport(null, FALSE, inserter = parent.get_creator())
```

Comment:
where is retrieved and sent actually set here?

Reply:
If I understand what you're asking, on the second input of the `doteleport()` proc. `TRUE` is send, `FALSE` is retrieve.